feat: Add Support for EKS Pod Identity

unitmatrix opened the pull request:
This pull request implements a new feature to add support for EKS Pod Identity as requested in issue #19 

#### Summary
This pull request introduces support for EKS Pod Identity in the module. The changes include the addition of new variables, resources, and documentation updates to enable the creation of Pod Identity roles and policies.

#### Changes
1. **Documentation Updates**:
   - Added instructions for enabling Pod Identity role and policy in README.md
   - Updated examples in README.md to include a new module configuration for Pod Identity.

2. **New Resources**:
   - Created eks_pod_identity.tf to define IAM resources for Pod Identity, including:
     - `aws_iam_policy.eks_pod_identity`
     - `aws_iam_role.eks_pod_identity`
     - `aws_iam_role_policy_attachment.eks_pod_identity`
     - `aws_iam_policy_document.eks_pod_identity_assume`
     - `aws_eks_pod_identity_association.this`

3. **Examples**:
   - Added a new example configuration `lbc_with_eks_pod_identity_role` in main.tf.

4. **Variables**:
   - Added new variables in variables.tf to support Pod Identity:
     - `eks_pod_identity_role_create`
     - `eks_pod_identity_role_name_prefix`
     - `eks_pod_identity_policy_enabled`
     - `eks_pod_identity_tags`

#### Impact
- These changes do not break existing functionality.
- Users can now enable Pod Identity support by setting the appropriate variables.

#### How to Test
- Follow the updated instructions in the README.md to enable Pod Identity.
- Use the new example configuration in main.tf to test the Pod Identity setup.

#### Notes
- Ensure that the AWS IAM permissions are correctly configured to allow the creation of the new IAM resources.

Closes #19 

Please review the changes and provide feedback.

jaygridley commented:
Hi @unitmatrix, adding support for EKS Pod Identity is a welcomed enhancement, but I would like to approach it without renaming existing IRSA related variables as it introduces unnecessary breaking changes. I would try to approach this as a separate functionality with own file, variables etc. Policy itself can be shared.

unitmatrix commented:
### Reply Message

Hi @jaygridley,

Thank you for your feedback. I have made the following changes to address your concerns:

1. **Separate Functionality**:
   - I have implemented the EKS Pod Identity support as a separate functionality with its own file (`iam_pod_identity.tf`) and variables.
   - The existing IRSA-related variables and functionality remain unchanged to avoid any breaking changes.

2. **Shared Policy**:
   - The policy for Pod Identity is defined separately but can be shared if needed.

I believe these changes align with your suggestions and maintain backward compatibility. Please review the updated implementation.

Thank you for your guidance.

Best regards.

jaygridley commented on variables.tf:
```diff
@@ -55,6 +55,30 @@ variable "service_account_name" {
 }
 
+variable "pod_identity_role_create" {
+  description = "Determines whether to enable support for EKS pod identity"
```
Lets keep the `description` as 3rd attribute of the variable

jaygridley commented on variables.tf:
```diff
@@ -55,6 +55,30 @@ variable "service_account_name" {
   description = "The k8s aws-loab-balancer-controller service account name"
 }
 
+variable "pod_identity_role_create" {
```
Maybe prefix with `eks_` would be more clear?

jaygridley commented on variables.tf:
```diff
@@ -55,6 +55,30 @@ variable "service_account_name" {
+variable "pod_identity_role_name_prefix" {
+  type        = string
+  default     = "lb-controller-pod-identity"
+  description = "The pod identity role name prefix for LB controller"
```
```suggestion
  description = "The EKS pod identity role name prefix for LB controller"
```

jaygridley commented on variables.tf:
```diff
@@ -55,6 +55,30 @@ variable "service_account_name" {
+variable "pod_identity_tags" {
+  type        = map(string)
+  default     = {}
+  description = "Pod identity resources tags"
```
```suggestion
  description = "The EKS pod identity resources tags"
```

jaygridley commented on iam_pod_identity.tf:
Rename to `eks_pod_identity.tf` for clarity?

jaygridley commented on iam_pod_identity.tf:
```diff
@@ -0,0 +1,47 @@
+locals {
+  pod_identity_role_create = var.enabled && var.service_account_create && var.pod_identity_role_create
```
```suggestion
  eks_pod_identity_role_create = var.enabled && var.service_account_create && var.eks_pod_identity_role_create
```

jaygridley replied:
I would add `eks_` prefix to all resources here as well

jaygridley commented on iam_pod_identity.tf:
```diff
@@ -0,0 +1,47 @@
+  path        = "/"
+  description = "Policy for aws-load-balancer-controller service"
+
+  policy = data.aws_iam_policy_document.this[0].json
```
We need to patch `data "aws_iam_policy_document" "this"` creation condition to
```
count = local.irsa_role_create && (var.irsa_policy_enabled || var.eks_pod_identity_policy_enabled) ? 1 : 0
```

jaygridley reviewed (changes requested):
Thanks for the changes. Looking good, but a few cosmetic touches would be nice ;-)

unitmatrix commented:
Thank you for the requested changes, @jaygridley. 

### Summary of Changes:
1. **Renamed Variables and Resources:**
   - Updated variable names and resource names to use the `eks_pod_identity` prefix instead of `pod_identity` for clarity and consistency.
   - Example: `pod_identity_role_create` is now `eks_pod_identity_role_create`.

2. **Updated Documentation:**
   - Modified the README and example files to reflect the new variable names.

3. **Logical Condition Update:**
   - Updated the logical condition for the creation of the IAM policy document to ensure that EKS pod identity can be implemented independently of the IRSA role.

Please review the changes and let me know if any further adjustments are needed.

jaygridley commented:
I think we just need to update https://github.com/lablabs/terraform-aws-eks-load-balancer-controller/pull/23#issue-2733447671 comment with latest changes in the variables a we are good to go. In the meantime I will try to test this out in our setup.

jaygridley commented:
I am considering one more thing by adding also `https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_pod_identity_association` into the module. I think it can help with the integration of EKS Pod identity further.

unitmatrix commented:
Hi @jaygridley ,

Thank you for the suggestion. I have added the `aws_eks_pod_identity_association` resource into the module and updated the README accordingly.

Best regards.

---

jaygridley reviewed (approved):
LGTM
